drivers/dma/dma_stm32u5.c

```diff
@@ -0,0 +1,757 @@
+		LOG_ERR("clock op failed\n");
+		return -EIO;
+	}
+
+	config->config_irq(dev);
+
+	for (uint32_t i = 0; i < config->max_streams; i++) {
+		config->streams[i].busy = false;
+	}
+
+	((struct dma_stm32_data *)dev->data)->dma_ctx.magic = 0;
+	((struct dma_stm32_data *)dev->data)->dma_ctx.dma_channels = 0;
+	((struct dma_stm32_data *)dev->data)->dma_ctx.atomic = 0;
+
+	return 0;
+}
+
+static int dma_stm32_get_status(const struct device *dev,
+				uint32_t id, struct dma_status *stat)
+{
+	const struct dma_stm32_config *config = dev->config;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+	struct dma_stm32_stream *stream;
+
+	/* Give channel from index 0 */
+	id = id - STM32_DMA_STREAM_OFFSET;
+	if (id >= config->max_streams) {
+		return -EINVAL;
+	}
+
+	stream = &config->streams[id];
+	stat->pending_length = LL_DMA_GetBlkDataLength(dma, dma_stm32_id_to_stream(id));
+	stat->dir = stream->direction;
+	stat->busy = stream->busy;
+
+	return 0;
+}
+
+static const struct dma_driver_api dma_funcs = {
+	.reload		 = dma_stm32_reload,
+	.config		 = dma_stm32_configure,
+	.start		 = dma_stm32_start,
+	.stop		 = dma_stm32_stop,
+	.get_status	 = dma_stm32_get_status,
+};
+
+#define DMA_STM32_OFFSET_INIT(index)
+#define DMA_STM32_MEM2MEM_INIT(index)
+
+#define DMA_STM32_INIT_DEV(index)					\
+static struct dma_stm32_stream						\
+	dma_stm32_streams_##index[DMA_STM32_##index##_STREAM_COUNT];	\
+									\
+const struct dma_stm32_config dma_stm32_config_##index = {		\
+	.pclken = { .bus = DT_INST_CLOCKS_CELL(index, bus),		\
+		    .enr = DT_INST_CLOCKS_CELL(index, bits) },		\
+	.config_irq = dma_stm32_config_irq_##index,			\
+	.base = DT_INST_REG_ADDR(index),				\
+	DMA_STM32_MEM2MEM_INIT(index)					\
+	.max_streams = DMA_STM32_##index##_STREAM_COUNT,		\
+	.streams = dma_stm32_streams_##index,				\
+	DMA_STM32_OFFSET_INIT(index)					\
+};									\
+									\
+static struct dma_stm32_data dma_stm32_data_##index = {			\
+};									\
+									\
+DEVICE_DT_INST_DEFINE(index,						\
+		    &dma_stm32_init,					\
+		    NULL,						\
+		    &dma_stm32_data_##index, &dma_stm32_config_##index,	\
+		    PRE_KERNEL_1, CONFIG_DMA_INIT_PRIORITY,		\
+		    &dma_funcs)
+
+#define DMA_STM32_DEFINE_IRQ_HANDLER(dma, chan)				\
+static void dma_stm32_irq_##dma##_##chan(const struct device *dev)	\
+{									\
+	dma_stm32_irq_handler(dev, chan);				\
+}
+
+
+#define DMA_STM32_IRQ_CONNECT(dma, chan)				\
+	do {								\
+		IRQ_CONNECT(DT_INST_IRQ_BY_IDX(dma, chan, irq),		\
+			    DT_INST_IRQ_BY_IDX(dma, chan, priority),	\
+			    dma_stm32_irq_##dma##_##chan,		\
+			    DEVICE_DT_INST_GET(dma), 0);		\
+		irq_enable(DT_INST_IRQ_BY_IDX(dma, chan, irq));		\
+	} while (0)
+
+/* STM32U5 soc has only one GPDMA instance of 15 channels */
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 0);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 3);
```

Comment:
Not blocking but use of `LISTIFY` should help reducing this (and make it conditional to a variable number if case happens)